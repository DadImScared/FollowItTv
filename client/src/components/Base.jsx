
import React, { Component } from 'react';

import _ from 'lodash';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';

import cyan from '@material-ui/core/colors/cyan';
import deepPurple from '@material-ui/core/colors/deepPurple';
import withWidth from '@material-ui/core/withWidth';
import { MuiThemeProvider, createMuiTheme, withStyles } from '@material-ui/core/styles';
import CssBaseline from '@material-ui/core/CssBaseline';

import Navbar from './Navbar';
import Navdrawer from './Navdrawer';
import Main from './Main';
import BottomNav from './BottomNav';

import { getFollowedShows } from '../actions/followedShows';
import styles from '../styles/Base.css';

const theme = createMuiTheme({
  palette: {
    primary: cyan,
    secondary: deepPurple
  }
});

export class  Base extends Component {

  constructor(...args) {
    super(...args);
    this.state = {
      isOpen: false,
      lastScrollTop: 0,
      directionDown: false,
      bottomOfPage: false
    };
  }

  async componentDidMount() {
    window.addEventListener('scroll', this.scrollListener);
    try {
      await getFollowedShows(this.props.dispatch);
    }
    catch ({ response: { data, status } }) {
      if (status === 401) {
        // do nothing if users token expired or is wrong
        return;
      }
      console.log(data);
      console.log(status);
    }
  }

  toggleNavdrawer = () => {
    this.setState((prevState => ({ isOpen: !prevState.isOpen })));
  };

  toggleHideNavBars = (scrollTop) => {
    const { lastScrollTop, directionDown, bottomOfPage } = this.state;
    const newState = {};
    if (scrollTop > lastScrollTop) { // scroll down
      if (!directionDown) {
        newState.directionDown = true;
      }
      if (this.atBottomOfPage() && !bottomOfPage) {
        newState.bottomOfPage = true;
      }
    }
    else { // scroll up
      if (directionDown) {
        newState.directionDown = false;
        newState.bottomOfPage = false;
      }
    }
    newState.lastScrollTop = scrollTop;
    this.setState(newState);
  };

  handleScroll = () => {
    const { width } = this.props;
    const scrollTop = this.getScrollTop();
    switch(width) {
      case 'xs':
      case 'sm':
      case 'md':
        this.toggleHideNavBars(scrollTop);
        break;
      default:
        return;
    }
  };

  scrollListener = _.throttle(this.handleScroll, 200);

  getScrollTop = () => {
    const docEl = document.documentElement;
    return (window.pageYOffset || docEl.scrollTop) - (docEl.clientTop || 0);
  };

  atBottomOfPage = () => {
    const isAtBottom = (window.innerHeight + window.pageYOffset) >= document.body.offsetHeight - 2;
    this.setState({ bottomOfPage: isAtBottom });
  };

  render() {
    const { classes, ...other } = this.props;
    const { isOpen, directionDown, bottomOfPage } = this.state;
    return (
      <MuiThemeProvider theme={theme}>
        <CssBaseline/>
        <div className={classes.appFrame}>
          <Navbar
            toggleNav={this.toggleNavdrawer}
            directionDown={directionDown}
          />
          <Navdrawer toggleNav={this.toggleNavdrawer} isOpen={isOpen} />
          <Main {...other} />
          <BottomNav
            navDrawerOpen={isOpen}
            toggleNavdrawer={this.toggleNavdrawer}
            {...other}
            directionDown={directionDown}
            bottomOfPage={bottomOfPage}
          />
        </div>
      </MuiThemeProvider>
    );
  }
}

export default withWidth()(withStyles(styles)(withRouter(connect()(Base))));
