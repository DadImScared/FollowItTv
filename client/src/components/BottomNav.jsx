
import React, { Component } from 'react';

import classNames from 'classnames';

import { withStyles } from '@material-ui/core/styles';
import BottomNavigation from '@material-ui/core/BottomNavigation';
import BottomNavigationAction from '@material-ui/core/BottomNavigationAction';
import Hidden from '@material-ui/core/Hidden';

import RestoreIcon from '@material-ui/icons/Restore';
import FavoriteIcon from '@material-ui/icons/Favorite';
import LocationOnIcon from '@material-ui/icons/LocationOn';

import styles from '../styles/BottomNavbar';


class BottomNav extends Component {
  constructor(...args) {
    super(...args);
    this.state = {
      value: 0,
      initialize: false
    };
  }

  handleChange = (event, value) => {
    this.setState({ value });
  };

  render() {
    const { classes, bottomOfPage, directionDown } = this.props;
    return (
      <Hidden mdUp>
        <BottomNavigation
          value={this.state.value}
          className={classNames({
            [classes.hideNavBar]: !bottomOfPage && directionDown,
            [classes.showNavBar]: !directionDown
          })}
          onChange={this.handleChange}
          showLabels
        >
          <BottomNavigationAction label="Recents" icon={<RestoreIcon />} />
          <BottomNavigationAction label="Favorites" icon={<FavoriteIcon />} />
          <BottomNavigationAction label="Nearby" icon={<LocationOnIcon />} />
        </BottomNavigation>
      </Hidden>
    );
  }
}

export default withStyles(styles)(BottomNav);
