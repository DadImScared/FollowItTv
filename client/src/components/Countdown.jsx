
import React, { Component } from 'react';

import moment from 'moment';

import Typography from 'material-ui/Typography';


export class Countdown extends Component {
  constructor(...args) {
    super(...args);
    this.state = {};
  }

  componentDidMount() {
    const { eventTime, startTime = moment(), callBack } = this.props;
    const diff = eventTime.diff(startTime);
    const duration = moment.duration(diff);
    if (duration.asMilliseconds() <= 0) {
      callBack && callBack();
      return;
    }
    this.setState({ duration });
    this.interval = setInterval(() => {
      this.updateTimer();
    }, 1000);
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  updateTimer = () => {
    const newDuration = this.state.duration.clone();
    newDuration.subtract(1, 'seconds');
    if (newDuration.asMilliseconds() <= 0) {
      const { callBack } = this.props;
      clearInterval(this.interval);
      callBack && callBack();
      return;
    }
    this.setState({ duration: newDuration });
  };

  getCountdown = () => {
    const { duration } = this.state;
    const days = duration.days();
    const hours = this.format(duration.hours());
    const minutes = this.format(duration.minutes());
    const seconds = this.format(duration.seconds());
    // don't show days if days are 0
    return `in ${days !== 0 ? `${days} days and `:''}${hours}:${minutes}:${seconds}`;
  };

  format = (time) => {
    const timeStr = time.toString();
    return timeStr.length === 1 ? (
      `0${timeStr}`
    ):(
      timeStr
    );
  };

  render() {
    return (
      <Typography>
        {
          this.state.duration ?
            this.getCountdown()
            :
            '00:00:'
        }
      </Typography>
    );
  }
}

export default Countdown;
