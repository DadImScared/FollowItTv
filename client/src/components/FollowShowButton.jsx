
import React, { Component } from 'react';

import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { withStyles } from 'material-ui/styles';
import Button from 'material-ui/Button';

import Heart from 'mdi-material-ui/Heart';
import HeartBroken from 'mdi-material-ui/HeartBroken';

import * as followShowActions from '../actions/followedShows';
import { addShow, getShow } from '../actions/shows';
import styles from '../styles/FollowShowButton.css';


export class FollowShowButton extends Component {
  constructor(...args) {
    super(...args);
    this.state = {
      isLoading: false
    };
  }

  handleClick = async () => {
    const { followedShows, showId, shows, followShow, unfollowShow } = this.props;
    if (this.state.isLoading) {
      return;
    }
    this.setState({ isLoading: true });
    if (!shows[showId]) {
      const show = await this.getAndUpdateShows(showId);
      shows[show.id] = show;
    }
    // follow or unfollow show
    const { schedule: { days } } = shows[showId];
    try {
      await followShowActions.postFollow(showId);
      if (followedShows.includes(showId)) {
        unfollowShow(days, showId);
      }
      else {
        followShow(days, showId);
      }
    }
    catch ({ response: { data, status } }) {
      console.log(status, data);
    }
    this.setState({ isLoading: false });
  };

  getAndUpdateShows = async (showId) => {
    try {
      const { data } = await getShow(showId);
      addShow(data.id, data);
      return data;
    }
    catch ({ response: { data, status } }) {
      console.log(data, status);
    }
  };

  render() {
    const { followedShows, showId, classes } = this.props;
    return (
      <Button onClick={this.handleClick}>
        <span className={classes.heartIcon}>
          { followedShows.includes(showId) ? <HeartBroken />:<Heart /> }
        </span>
        { followedShows.includes(showId) ? 'Un follow':'Follow'}
      </Button>
    );
  }
}

const mapStateToProps = ({ shows, followedShowsById }) => ({
  shows,
  followedShows: followedShowsById
});

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators({
    followShow: followShowActions.followShow,
    unfollowShow: followShowActions.unfollowShow,
    addShow
  }, dispatch);
};

export default withStyles(styles)(connect(mapStateToProps, mapDispatchToProps)(FollowShowButton));
