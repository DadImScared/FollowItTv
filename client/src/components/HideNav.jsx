
import  React from 'react';
import PropTypes from 'prop-types';

import AppBar from '@material-ui/core/AppBar';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';

import styles from '../styles/HideNav.css';

const HideNav = ({ classes, directionDown, children }) => (
  <AppBar
    className={classNames(classes.appBar, {
      [classes.stickyDown]: directionDown,
      [classes.stickyUp]: !directionDown
    })}
    color={'secondary'}
    position={'sticky'}
  >
    {children}
  </AppBar>
);

HideNav.propTypes = {
  classes: PropTypes.shape({
    appBar: PropTypes.string.isRequired,
    stickyDown: PropTypes.string.isRequired,
    stickyUp: PropTypes.string.isRequired
  }).isRequired,
  directionDown: PropTypes.bool.isRequired
};

export default withStyles(styles)(HideNav);
