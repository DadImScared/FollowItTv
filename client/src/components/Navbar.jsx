
import React from 'react';

import { withStyles } from 'material-ui/styles';
import AppBar from 'material-ui/AppBar';
import Toolbar from 'material-ui/Toolbar';
import Hidden from 'material-ui/Hidden';
import IconButton from 'material-ui/IconButton';

import MenuIcon from 'material-ui-icons/Menu';

import styles from '../styles/Navbar.css';

const Navbar = ({ classes, toggleNav }) => (
  <AppBar color={'secondary'} className={classes.appBar}>
    <Toolbar>
      <Hidden mdUp>
        <IconButton color={'inherit'} onClick={toggleNav}>
          <MenuIcon />
        </IconButton>
      </Hidden>
    </Toolbar>
  </AppBar>
);

export default withStyles(styles)(Navbar);
