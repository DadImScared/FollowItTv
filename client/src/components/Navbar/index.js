
import React from 'react';
import PropTypes from 'prop-types';

import classNames from 'classnames';
import { connect } from 'react-redux';
import Cookies from 'js-cookie';
import { Link, withRouter } from 'react-router-dom';

import { withStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Hidden from '@material-ui/core/Hidden';
import Button from '@material-ui/core/Button';

import { logOut } from '../../actions/users';
import LoggedInButtons from './LoggedInButtons';
import LoggedOutButtons from './LoggedOutButtons';
import MoreOptions from './MoreOptions';
import Searchbar from './Searchbar';
import styles from '../../styles/Navbar.css';

export const Navbar = ({ classes, loggedIn, logOut, directionDown, ...other }) => {
  const logOutUser = () => {
    Cookies.remove('token');
    logOut();
  };

  return (
    <AppBar
      color={'secondary'}
      className={classes.appBar}
      classes={{
        root: classNames({
          [classes.hideNavBar]: directionDown,
          [classes.showNavBar]: !directionDown
        })
      }}
    >
      <Toolbar style={{ justifyContent: 'space-between' }}>
        <Hidden smUp>
          <div />
        </Hidden>
        <div>
          <Searchbar {...other} />
        </div>
        <Hidden smDown>
          <div style={{ display: 'flex' }}>
            <Button component={Link} to={'/schedule'}>
              Schedule
            </Button>
            {
              loggedIn ?
                <LoggedInButtons logOut={logOutUser} />
                :
                <LoggedOutButtons/>
            }
          </div>
        </Hidden>
        <Hidden mdUp>
          <div>
            <MoreOptions {...other} logOut={logOutUser} loggedIn={loggedIn} />
          </div>
        </Hidden>
      </Toolbar>
    </AppBar>
  );
};

Navbar.propTypes = {
  classes: PropTypes.shape({
    appBar: PropTypes.string.isRequired,
    hideNavBar: PropTypes.string.isRequired,
    showNavBar: PropTypes.string.isRequired
  }).isRequired,
  loggedIn: PropTypes.bool.isRequired,
  directionDown: PropTypes.bool.isRequired,
  logOut: PropTypes.func.isRequired
};

const mapStateToProps = ({ users: { loggedIn } }) => ({
  loggedIn
});

const mapDispatchToProps = (dispatch) => ({
  logOut: () => dispatch(logOut())
});

export default withStyles(styles)(withRouter(connect(mapStateToProps, mapDispatchToProps)(Navbar)));
