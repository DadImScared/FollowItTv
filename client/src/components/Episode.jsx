
import React, { Component } from 'react';

import { Link } from 'react-router-dom';

import classnames from 'classnames';
import { withStyles } from 'material-ui/styles';
import Button from 'material-ui/Button';
import Card, { CardHeader, CardMedia, CardContent, CardActions } from 'material-ui/Card';
import Typography from 'material-ui/Typography';
import Collapse from 'material-ui/transitions/Collapse';
import IconButton from 'material-ui/IconButton';
import SvgIcon from 'material-ui/SvgIcon';

import Television from 'mdi-material-ui/Television';
import ExpandMoreIcon from 'material-ui-icons/ExpandMore';

import Summary from './Summary';
import FollowShowButton from './FollowShowButton';
import styles from '../styles/Episode.css';


export class Episode extends Component {
  constructor(...args) {
    super(...args);
    this.state = {
      isOpen: false,
      height: 0
    };
  }

  handleExpand = () => {
    this.setState({ isOpen: !this.state.isOpen });
  };

  setRef = (el) => {
    this.el = el;
  };

  componentDidMount() {
    if (this.el) {
      this.setState({ height: this.el.clientHeight });
    }
  }

  render() {
    const {
      item: {
        name,
        show,
        airtime,
        airdate,
        summary,
        image
      } = {},
      shows = {},
      showActions = true,
      classes,
      followProps = {}
    } = this.props;
    const currentShow = this.props.show || shows[show];
    const currentId = (this.props.show && this.props.show.id) || show;
    return (
      <div>
        <Card>
          <CardHeader
            title={
              <div style={{ display: 'flex', justifyContent: 'space-between', flexWrap: 'wrap', alignItems: 'center' }}>
                <Typography variant={'subheading'}>
                  {this.props.show ? name:currentShow.name || 'no name'}
                </Typography>
                <div style={{ display: 'flex', alignItems: 'center' }}>
                  <Typography style={{ marginRight: '4px' }}>
                    {currentShow.network && currentShow.network.name}
                  </Typography>
                  <a href={`${currentShow.officialSite}`} style={{ display: 'flex' }}>
                    <SvgIcon color={'primary'}>
                      <Television />
                    </SvgIcon>
                  </a>
                </div>
              </div>
            }
            subheader={`${this.props.show ? `${airdate} | `:''}${airtime}${!this.props.show ? ` | ${name}`:''}`}
          />
          <CardMedia
            component={'img'}
            className={classes.media}
            // image={shows[show] && shows[show].image && shows[show].image.medium}
            image={(image && image.medium) || (currentShow.image && currentShow.image.medium)}
            title={`episode ${name}. show ${currentShow.name || 'no name'}`}
          />
          <CardContent>
            <Collapse collapsedHeight={'40px'} in={this.state.isOpen}>
              <CardContent>
                <Typography component={'div'}>
                  {
                    summary ?
                      <Summary summary={summary} setRef={this.setRef} />
                      :
                      'No summary entered.'
                  }
                </Typography>
              </CardContent>
            </Collapse>
            <div
              id={'read-more'}
              style={{
                display: 'flex',
                justifyContent: 'space-between',
                visibility: this.state.height >= 40 ? 'visible':'hidden'
              }}>
              <Typography style={{ alignSelf: 'center' }}>read more</Typography>
              <IconButton
                color={'secondary'}
                className={classnames(classes.expand, {
                  [classes.expandOpen]: this.state.isOpen
                })}
                onClick={this.handleExpand}
              >
                <ExpandMoreIcon />
              </IconButton>
            </div>
          </CardContent>
          {
            showActions ?
              <CardActions>
                <div style={{ display: 'flex', justifyContent: 'space-between', width: '100%' }}>
                  <FollowShowButton {...followProps} showId={currentId} />
                  <Button component={Link} to={`/show/${currentId}/general`} color={'secondary'}>
                    View show
                  </Button>
                </div>
              </CardActions>
              :
              null
          }
        </Card>
      </div>
    );
  }
}

export default withStyles(styles)(Episode);
