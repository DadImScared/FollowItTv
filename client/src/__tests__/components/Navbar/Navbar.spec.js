
import React from 'react';

import sinon from 'sinon';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { Navbar } from '../../../components/Navbar';

describe('Navbar', () => {
  let props;

  beforeEach(() => {
    props = {
      classes: {
        appBar: 'appbar',
        hideNavBar: 'hideNav',
        showNavBar: 'showNav'
      },
      loggedIn: true,
      directionDown: true,
      logOut: sinon.spy(),
      history: {
        push: sinon.spy()
      }
    };
  });
  
  it('should render logged in options if loggedIn', () => {
    const wrapper = shallow(<Navbar {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render logged out options if loggedIn = false', () => {
    const wrapper = shallow(<Navbar {...{ ...props, loggedIn: false }} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
