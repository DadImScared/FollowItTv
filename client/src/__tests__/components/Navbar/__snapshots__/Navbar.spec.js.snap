// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar should render logged in options if loggedIn 1`] = `
<WithStyles(AppBar)
  className="appbar"
  classes={
    Object {
      "root": "hideNav",
    }
  }
  color="secondary"
>
  <WithStyles(Toolbar)
    style={
      Object {
        "justifyContent": "space-between",
      }
    }
  >
    <Hidden
      implementation="js"
      lgDown={false}
      lgUp={false}
      mdDown={false}
      mdUp={false}
      smDown={false}
      smUp={true}
      xlDown={false}
      xlUp={false}
      xsDown={false}
      xsUp={false}
    >
      <div />
    </Hidden>
    <div>
      <Searchbar
        history={
          Object {
            "push": [Function],
          }
        }
      />
    </div>
    <Hidden
      implementation="js"
      lgDown={false}
      lgUp={false}
      mdDown={false}
      mdUp={false}
      smDown={true}
      smUp={false}
      xlDown={false}
      xlUp={false}
      xsDown={false}
      xsUp={false}
    >
      <div
        style={
          Object {
            "display": "flex",
          }
        }
      >
        <WithStyles(Button)
          component={[Function]}
          to="/schedule"
        >
          Schedule
        </WithStyles(Button)>
        <LoggedInButtons
          logOut={[Function]}
        />
      </div>
    </Hidden>
    <Hidden
      implementation="js"
      lgDown={false}
      lgUp={false}
      mdDown={false}
      mdUp={true}
      smDown={false}
      smUp={false}
      xlDown={false}
      xlUp={false}
      xsDown={false}
      xsUp={false}
    >
      <div>
        <MoreOptions
          history={
            Object {
              "push": [Function],
            }
          }
          logOut={[Function]}
          loggedIn={true}
        />
      </div>
    </Hidden>
  </WithStyles(Toolbar)>
</WithStyles(AppBar)>
`;

exports[`Navbar should render logged out options if loggedIn = false 1`] = `
<WithStyles(AppBar)
  className="appbar"
  classes={
    Object {
      "root": "hideNav",
    }
  }
  color="secondary"
>
  <WithStyles(Toolbar)
    style={
      Object {
        "justifyContent": "space-between",
      }
    }
  >
    <Hidden
      implementation="js"
      lgDown={false}
      lgUp={false}
      mdDown={false}
      mdUp={false}
      smDown={false}
      smUp={true}
      xlDown={false}
      xlUp={false}
      xsDown={false}
      xsUp={false}
    >
      <div />
    </Hidden>
    <div>
      <Searchbar
        history={
          Object {
            "push": [Function],
          }
        }
      />
    </div>
    <Hidden
      implementation="js"
      lgDown={false}
      lgUp={false}
      mdDown={false}
      mdUp={false}
      smDown={true}
      smUp={false}
      xlDown={false}
      xlUp={false}
      xsDown={false}
      xsUp={false}
    >
      <div
        style={
          Object {
            "display": "flex",
          }
        }
      >
        <WithStyles(Button)
          component={[Function]}
          to="/schedule"
        >
          Schedule
        </WithStyles(Button)>
        <LoggedOutButtons />
      </div>
    </Hidden>
    <Hidden
      implementation="js"
      lgDown={false}
      lgUp={false}
      mdDown={false}
      mdUp={true}
      smDown={false}
      smUp={false}
      xlDown={false}
      xlUp={false}
      xsDown={false}
      xsUp={false}
    >
      <div>
        <MoreOptions
          history={
            Object {
              "push": [Function],
            }
          }
          logOut={[Function]}
          loggedIn={false}
        />
      </div>
    </Hidden>
  </WithStyles(Toolbar)>
</WithStyles(AppBar)>
`;
